utils/signer/src/lib.rs

```diff
@@ -187,38 +194,549 @@ impl Signer {
+                anyhow::bail!(
+                    "{} environment variable contains invalid unicode: {:?}",
+                    var_name,
+                    os_str
+                )
+            }
+        }
+    }
+}
+
+/// Configuration for adaptive gas pricing strategy.
+///
+/// This config controls how gas prices are calculated when network gas prices
+/// exceed a threshold. When gas is above the threshold, the filler uses
+/// historical average baseFee (from `eth_feeHistory`) multiplied by a factor,
+/// with gradual increases after a timeout.
+#[derive(Clone, Debug)]
+pub struct AdaptiveGasConfig {
+    /// Threshold above which to apply historical smoothing (wei).
+    /// Default: 3 gwei (3_000_000_000 wei)
+    pub gas_price_threshold: u128,
+    /// Number of blocks for history (~900 for 3 hours at 12s/block).
+    /// Default: 900
+    pub history_blocks: u64,
+    /// Multiplier for historical average (e.g., 1.1 = 110%).
+    /// Default: 1.1
+    pub price_multiplier: f64,
+    /// Timeout in minutes before gradual increase starts.
+    /// Default: 30
+    pub fallback_timeout_minutes: u64,
+    /// Percentage increase per timeout period.
+    /// Default: 10
+    pub fallback_increase_percent: u64,
+}
+
+impl Default for AdaptiveGasConfig {
+    fn default() -> Self {
+        Self {
+            gas_price_threshold: 3_000_000_000,   // 3 gwei
+            history_blocks: 900,                   // ~3 hours at 12s/block
+            price_multiplier: 1.1,                 // 110%
+            fallback_timeout_minutes: 30,
+            fallback_increase_percent: 10,
+        }
+    }
+}
+
+impl AdaptiveGasConfig {
+    /// Creates a new AdaptiveGasConfig from environment variables.
+    ///
+    /// All values are optional and fall back to defaults if not set.
+    /// Returns an error if an environment variable is set but contains an invalid value.
+    pub fn from_env() -> Result<Self> {
+        let mut config = Self::default();
+
+        if let Some(threshold) = Self::parse_env_var_u128("GAS_PRICE_THRESHOLD")? {
+            config.gas_price_threshold = threshold;
+        }
+        if let Some(blocks) = Self::parse_env_var_u64("GAS_HISTORY_BLOCKS")? {
+            config.history_blocks = blocks;
+        }
+        if let Some(multiplier) = Self::parse_env_var_f64("GAS_PRICE_MULTIPLIER")? {
+            if multiplier <= 0.0 {
+                anyhow::bail!("GAS_PRICE_MULTIPLIER must be positive, got: {}", multiplier);
+            }
+            config.price_multiplier = multiplier;
+        }
+        if let Some(timeout) = Self::parse_env_var_u64("GAS_FALLBACK_TIMEOUT_MINUTES")? {
+            config.fallback_timeout_minutes = timeout;
+        }
+        if let Some(percent) = Self::parse_env_var_u64("GAS_FALLBACK_INCREASE_PERCENT")? {
+            config.fallback_increase_percent = percent;
+        }
+
+        Ok(config)
+    }
+
+    fn parse_env_var_u128(var_name: &str) -> Result<Option<u128>> {
```

Comment:
This could be done with some generics, to avoid repetition. Also the parse_env that is used above for the max_fee_per_gas and the priority one